lint/checkstyle.bzl

```diff
@@ -0,0 +1,149 @@
+    runtime_deps = ["@com_puppycrawl_tools_checkstyle"],
+)
+```
+
+Finally, declare an aspect for it, typically in `tools/lint/linters.bzl`:
+
+```starlark
+load("@aspect_rules_lint//lint:checkstyle.bzl", "checkstyle_aspect")
+
+checkstyle = checkstyle_aspect(
+    binary = "@@//tools/lint:checkstyle",
+    config = ["@@//:checkstyle.xml"],
+)
+```
+"""
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_jar")
+load("//lint/private:lint_aspect.bzl", "LintOptionsInfo", "filter_srcs", "noop_lint_action", "output_files", "should_visit")
+
+_MNEMONIC = "AspectRulesLintCheckstyle"
+
+def checkstyle_action(ctx, executable, srcs, config, data, stdout, exit_code = None, options = []):
+    """Run Checkstyle as an action under Bazel.
+
+    Based on https://checkstyle.sourceforge.io/cmdline.html
+
+    Args:
+        ctx: Bazel Rule or Aspect evaluation context
+        executable: label of the the Checkstyle program
+        srcs: java files to be linted
+        config: label of the checkstyle.xml file
+        data: labels of additional xml files such as suppressions.xml
+        stdout: output file to generate
+        exit_code: output file to write the exit code.
+            If None, then fail the build when Checkstyle exits non-zero.
+        options: additional command-line options, see https://checkstyle.sourceforge.io/cmdline.html
+    """
+    inputs = srcs + [config] + data
+    outputs = [stdout]
+
+    # Wire command-line options, see
+    # https://checkstyle.sourceforge.io/cmdline.html
+    args = ctx.actions.args()
+    args.add_all(options)
+
+    args.add_all(["-c", config.path])
+    args.add_all(srcs)
+
+    if exit_code:
+        command = "{CHECKSTYLE} $@ >{stdout}; echo $? > " + exit_code.path
+        outputs.append(exit_code)
+    else:
+        # Create empty stdout file on success, as Bazel expects one
+        command = "{CHECKSTYLE} $@ && touch {stdout}"
+
+    ctx.actions.run_shell(
+        inputs = inputs,
+        outputs = outputs,
+        command = command.format(CHECKSTYLE = executable.path, stdout = stdout.path),
+        arguments = [args],
+        mnemonic = _MNEMONIC,
+        tools = [executable],
+        progress_message = "Linting %{label} with Checkstyle",
+    )
+
+# buildifier: disable=function-docstring
+def _checkstyle_aspect_impl(target, ctx):
+    if not should_visit(ctx.rule, ctx.attr._rule_kinds):
+        return []
+
+    files_to_lint = filter_srcs(ctx.rule)
+    outputs, info = output_files(_MNEMONIC, target, ctx)
+    if len(files_to_lint) == 0:
+        noop_lint_action(ctx, outputs)
+        return [info]
+
+    format_options = []
+    checkstyle_action(ctx, ctx.executable._checkstyle, files_to_lint, ctx.file._config, ctx.files._data, outputs.human.out, outputs.human.exit_code, format_options)
+    checkstyle_action(ctx, ctx.executable._checkstyle, files_to_lint, ctx.file._config, ctx.files._data, outputs.machine.out, outputs.machine.exit_code)
```

Comment:
for machine readable output, maybe you want the sarif output format? 

> -f format - Specifies the output format. Valid values: xml, sarif, plain for [XMLLogger](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/XMLLogger.html), [SarifLogger](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/SarifLogger.html), and [DefaultLogger](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/DefaultLogger.html) respectively. Defaults to plain.

Reply:
Done! It's not clear to me how the machine output gets used.